Clean/Utils/OfflineMemory.lean

```diff
@@ -0,0 +1,613 @@
+-/
+theorem MemoryAccessList.isConsistent_iff_all_single_address (accesses : MemoryAccessList) (h_sorted : accesses.isTimestampSorted) :
+    MemoryAccessList.isConsistentOnline accesses h_sorted ↔
+    (∀ addr : ℕ, MemoryAccessList.isConsistentSingleAddress (MemoryAccessList.filterAddress accesses addr) (MemoryAccessList.filterAddress_sorted accesses h_sorted addr)) := by
+  constructor
+  · intro h addr
+    have h' := MemoryAccessList.isConsistentSingleAddress_iff (accesses.filterAddress addr) addr (MemoryAccessList.filterAddress_sorted accesses h_sorted addr)
+        (by simp only [filterAddress, List.all_filter, Bool.not_or_self, List.all_eq_true,
+          implies_true])
+    exact h'.mp (MemoryAccessList.isConsistentOnline_filter_of_consistentOnline accesses h_sorted h addr)
+  · intro h
+    -- by induction on the list of accesses
+    induction accesses with
+    | nil =>
+      simp [isConsistentOnline]
+    | cons head tail ih =>
+      obtain ⟨t, a, r, w⟩ := head
+      have h_sorted' : isTimestampSorted tail := by
+        unfold isTimestampSorted at h_sorted
+        exact List.Sorted.of_cons h_sorted
+      specialize ih h_sorted'
+      have h_tail := MemoryAccessList.isConsistentSingleAddress_cons_forall (t, a, r, w) tail h_sorted h
+      specialize ih h_tail
+      simp only [isConsistentOnline, ih, and_true]
+
+      have h_tail_filter_sorted : (MemoryAccessList.filterAddress tail a).isTimestampSorted := by
+        simp only [filterAddress]
+        apply List.Sorted.filter
+        exact h_sorted'
+
+      have h_filtered_sorted : MemoryAccessList.isTimestampSorted ((t, a, r, w) :: (MemoryAccessList.filterAddress tail a)) := by
+        simp only [isTimestampSorted, List.sorted_cons, filterAddress, List.mem_filter,
+          and_imp] at ⊢ h_sorted
+        simp_all only [implies_true, true_and]
+        apply List.Sorted.filter
+        assumption
+
+
+      have h' := MemoryAccessList.lastWriteValue_filter tail h_sorted' a h_tail_filter_sorted
+
+      specialize h a
+      simp [filterAddress_cons] at h
+      -- rw [h']
+
+      have h_iff := MemoryAccessList.isConsistentSingleAddress_iff ((t, a, r, w) :: filterAddress tail a) a h_filtered_sorted
+      simp only [filterAddress, List.all_cons, decide_true, List.all_filter, Bool.not_or_self,
+        Bool.true_and, List.all_eq_true, implies_true, forall_const] at h_iff
+      have is_consistent_online_full := h_iff.mpr h
+      simp only [isConsistentOnline] at is_consistent_online_full
+      simp only [is_consistent_online_full.left]
+      rw [eq_comm]
+      apply MemoryAccessList.lastWriteValue_filter
+
+/--
+  Offline version of memory consistency checking.
+-/
+def MemoryAccessList.isConsistentOffline (accesses : MemoryAccessList) (h_sorted : accesses.isAddressTimestampSorted) : Prop := match accesses with
+  | [] => True -- no memory access is trivially consistent
+  | (_timestamp, _addr, readValue, _writeValue) :: [] => readValue = 0
```

Comment:
This line is actually looking at the last access, usually not the first. So I have a feeling something else is wanted.

Reply:
never mind, I wasn't aware the sort is reversed.